commit 35025735a79eaa894c43837b94fd33c9d6b122df
Author: David Woodhouse <dwmw@amazon.co.uk>
Date:   Thu Mar 3 15:41:18 2022 +0000

    KVM: x86/xen: Support direct injection of event channel events
    
    This adds a KVM_XEN_HVM_EVTCHN_SEND ioctl which allows direct injection
    of events given an explicit { vcpu, port, priority } in precisely the
    same form that those fields are given in the IRQ routing table.
    
    Userspace is currently able to inject 2-level events purely by setting
    the bits in the shared_info and vcpu_info, but FIFO event channels are
    harder to deal with; we will need the kernel to take sole ownership of
    delivery when we support those.
    
    A patch advertising this feature with a new bit in the KVM_CAP_XEN_HVM
    ioctl will be added in a subsequent patch.
    
    Signed-off-by: David Woodhouse <dwmw@amazon.co.uk>
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
    Message-Id: <20220303154127.202856-9-dwmw2@infradead.org>
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

diff --git a/include/uapi/linux/kvm.h b/include/uapi/linux/kvm.h
index 91a6fe4e02c0..49cd2e9e0f6a 100644
--- a/include/uapi/linux/kvm.h
+++ b/include/uapi/linux/kvm.h
@@ -1699,6 +1699,9 @@ struct kvm_xen_hvm_attr {
 #define KVM_XEN_VCPU_GET_ATTR	_IOWR(KVMIO, 0xca, struct kvm_xen_vcpu_attr)
 #define KVM_XEN_VCPU_SET_ATTR	_IOW(KVMIO,  0xcb, struct kvm_xen_vcpu_attr)
 
+/* Available with KVM_CAP_XEN_HVM / KVM_XEN_HVM_CONFIG_EVTCHN_SEND */
+#define KVM_XEN_HVM_EVTCHN_SEND	_IOW(KVMIO,  0xd0, struct kvm_irq_routing_xen_evtchn)
+
 #define KVM_GET_SREGS2             _IOR(KVMIO,  0xcc, struct kvm_sregs2)
 #define KVM_SET_SREGS2             _IOW(KVMIO,  0xcd, struct kvm_sregs2)